### DISCUSSION TITLE
Precompiling and bundling dependencies with fatal bugs inhibits upgrades of next.js

### DISCUSSION THREAD

**Discussion by cdaringe** [vercel/next.js#38450] [Ideas]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

not applicable

### What browser are you using? (if relevant)

n/a

### How are you deploying your application? (if relevant)

n/a

### Describe the Bug

Next.js is taking the position of bundling strategic dependencies for packaging performance. One such example of this is `node-fetch`, which can be seen in the compiled section of the next.js npm tarball.

`node-fetch` _generally works._ However, in our project, we have 10s of teams issues 1000s of queries. With that volume, probabilistically, we start hitting all of the edge cases. `node-fetch` has many open issues (https://github.com/node-fetch/node-fetch/issues), some quite fatal [1](https://github.com/node-fetch/node-fetch/issues/1568) [2](https://github.com/node-fetch/node-fetch/issues/1380). By using package manager solutions (overrides, resolutions, hooks, etc), we were able to circumvent flaws in transitive dependencies easily before, to use versions and patches to workaround issues of interest gracefully. However, precompiling such dependencies in the distribution makes correcting such issues ergonomically challenging. Next.js ejects from traditional dependency distribution, and even [meddles with node internals (module resolution)](https://github.com/vercel/next.js/pull/32742/files#diff-f5e0118d3ce0589ca0de992c92d52aa5bd8014dd4a2d95fb8ff0b6ee153d0d43R132-R143), hindering me from using a specific version of node-fetch that I require in the active node process, both within a next.js context & within my custom server.

Possible solutions:

1. Fix the upstream bugs, then upgrade next.js to use the latest upstream versions. This is the ideal outcome, always. However, in a world of perpetual change, it's not the most tenable. Bugs will always exist, and consuming projects will always lag behind.
1. Use patch-package! This could work, however, next.js code would need to be deeply modified, as various imports are now  explicitly too local `./dist/node-fetch`, vs just `node-fetch`. next.js now has hard coupling to its own vendored 3p dependencies, so a patch would not be simple or precise, it would impact many modules, thus be risky/brittle (as with patch-package, generally).
1. use normal node/npm dependency distribution

### Expected Behavior

Various behaviors could be expected, and are all debatable/subjective.

- To sustain prior art, use normal node packaging-isms for dependencies, that allows custom resolutions as offered by package managers.
  - Perhaps make tradeoffs. For simple/stable dependencies, keep w/ the precompile vendoring strategy. But for unstable (subjective) and problematic dependencies such as something as _core to web as fetch_, choose to not bundle such dependencies.
- Allow users to swap in dependencies. next.js is meddling with `module` resolution to point all imports of `node-fetch` to it's compiled version. This is very problematic, especially given that we use a custom server and do all sorts of stuff with fetch outside of `next.js`. We cannot upgrade to next@latest because of this change.

My ideal outcome would have `node-fetch` removed from https://github.com/vercel/next.js/pull/32742's changeset (cc/ @ijjk , as they may have stronger opinions on this as the implementer of 32742)

### Link to reproduction

n/a

### To Reproduce

n/a, this is a design discussion

**Comment by balazsorban44** [comment]
Are you using a custom server by chance? In that case, `fetch` is only polyfilled if your environment doesn't already have it. (Also not polyfilled if you are using Node.js 18+ which comes with `fetch` out of the box)

https://github.com/vercel/next.js/blob/canary/packages/next/server/node-polyfill-fetch.js#L8

**Reply by cdaringe** [reply]
- yes, i do use a custom server
- I believe that next.js uses its `node-fetch` in dev all of the time by means of [module resolution interception](https://github.com/vercel/next.js/pull/32742/files#diff-f5e0118d3ce0589ca0de992c92d52aa5bd8014dd4a2d95fb8ff0b6ee153d0d43R128-R143)

There are direct imports directly to this bundled module in next's source code as well. Thus, there are multiple avenues for fetch invocations to use next's bundled version, even if you are a) not using a custom server, and b) using a node version with global.fetch available.
